### PR TITLE
[Android][expo-barcode-scanner] update barcode scanner library in Android

### LINKED ISSUE #21995: expo-camera can't read QR code containing “ ’ in the Android Expo Go app
### Minimal reproducible example

https://github.com/toshiyuki-suzuki-yukashikado/my-expo

### Summary

When trying to read a QR code containing “ ’ using expo-camera in the Android Expo Go app, it always fails. By failure, we mean that when the problematic QR code is captured by the camera, nothing happens. The recognized QR code fails internally within Android before reaching react-native.
The react-native log does not output anything. The adb device log outputs the followingadb shell
```
2023-04-04 15:40:30.332 14796-14875 GMVBarCodeScanner host.exp.exponent E Failed to detect barcode: Attempt to invoke virtual method 'int java.lang.String.hashCode()' on a null object reference
```

~The relevant part of this error message is probably located [here](https://github.com/expo/expo/blob/33efd9c7bdc89dccfe8a77d06e6c3e4fd693adde/android/versioned-abis/expoview-abi48_0_0/src/main/java/abi48_0_0/expo/modules/camera/ExpoCameraView.kt#L74)~
This  error is caused by com.google.android.gms:play-services-vision:19.0.0. This error is also caused by'com.google.android.gms:play-services-vision:20.1.3.

This issue does not occur on iOS and is only a problem on Android. Additionally, the same issue occurs with the expo-barcode-scanner as well.

## npx expo-env-info
```
expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.6
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 16.13.1 - ~/.anyenv/envs/nodenv/versions/16.13.1/bin/node
      Yarn: 1.22.17 - ~/.anyenv/envs/nodenv/versions/16.13.1/bin/yarn
      npm: 8.1.2 - ~/.anyenv/envs/nodenv/versions/16.13.1/bin/npm
      Watchman: 2023.02.27.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /Users/toshiyuki/.anyenv/envs/rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    IDEs:
      Android Studio: Electric Eel 2022.1.1 Patch 2 Electric Eel 2022.1.1 Patch 2
      Xcode: 14.2/14C18 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~48.0.10 => 48.0.10 
      react: 18.2.0 => 18.2.0 
      react-native: 0.71.6 => 0.71.6 
    npmGlobalPackages:
      expo-cli: 6.0.5
    Expo Workflow: managed
```

## reproduction environment
You can reproduce the issue by using the following command and source code, or by cloning [this code](https://github.com/toshiyuki-suzuki-yukashikado/my-expo)
1. setup
```
npx create-expo-app --template
(template select: blank typescript)
(app name: my-expo)
cd my-expo
```
2.  install library
```
npx expo install expo-camera

```
3. edit app.tsx
``` typescript
// App.tsx

import { StatusBar } from 'expo-status-bar';
import { StyleSheet, View } from 'react-native';
import {useEffect, useState} from 'react';
import {Camera, CameraType} from "expo-camera";

export default function App() {
  const [permission, setPermission] = useState<{
    isGranted: boolean
    isChecked: boolean
  }>({
    isGranted: false,
    isChecked: false
  })
  useEffect(() => {
    Camera.getCameraPermissionsAsync().then(({status}) => {
      setPermission({
        isGranted: status === 'granted',
        isChecked: true
      })
    })
  }, [])
  useEffect(() => {
    if (permission.isChecked && !permission.isGranted) {
      Camera.requestCameraPermissionsAsync().then(({status}) => {
        setPermission({
          isGranted: status === 'granted',
          isChecked: true
        })
      })
    }
  }, [permission])
  return (
    <View style={styles.container}>
     {permission.isGranted && (
        <Camera
          type={CameraType.back}
          ratio={'16:9'}
          style={styles.camera}
          onBarCodeScanned={(scannerResult)=>{
            console.log(scannerResult)
          }} />
       )}
      <StatusBar style="auto" />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  camera: {
    flex: 1,
  }
});
```
4. run
```
npm run start 
```
5. read qrcode
The QR code below is generated from the text "test’". I believe you can observe the same issue if you try to read it.

qr20230405152754467

### PR DESCRIPTION
**Pull request by toshiyuki-suzuki-yukashikado**
# Why
Until now, we were using 'com.google.android.gms:play-services-vision' for QR code scanning of expo-barcode-scanner in Android. However, it is deprecated and has not been updated since 2020. We have now switched to using the recommended project, 'com.google.mlkit:barcode-scanning'. Additionally, a bug that prevented QR codes with special characters from being read has been fixed.

Resolve: #21995

# How

We've replaced the `com.google.android.gms:play-services-vision` (gms) library with `com.google.mlkit:barcode-scanning` (mlkit), keeping the interface as intact as possible. MLKit is the successor of GMS and supports the same types of barcodes, so no features are lost. While MLKit only supports asynchronous barcode scanning, we've achieved synchronous processing using lightweight Kotlin coroutines.

The issue with QR codes containing specific characters likely occurs  when converting from `barcode.rawBytes` to `barcode.rawValue` within MLKit. As a result, `barcode.rawValue` becomes `null`. To address this, we use `String(barcode.rawBytes)` as a new value when barcode.rawValue is null and barcode.rawBytes has a value.

# Test Plan
bare-expo
- [x] can read barcode 128
- [x] can read qrcord include special characters (but garbled characters occur)

https://user-images.githubusercontent.com/95330519/231669557-1b56d248-8a6c-4173-89b2-ebd5934d4aad.mp4

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

### CODE REVIEW

**Comment by toshiyuki-suzuki-yukashikado** [comment]
@lukmccall 
Sorry for being so busy. I would like to know if there is any test, documentation, etc. needed for you to review. I would be happy to get a response!

**Comment by lukmccall** [comment]
Hello @toshiyuki-suzuki-yukashikado, thank you for your contribution. My apologies for the delayed response, as I have been quite occupied lately. However, I will make it a priority to review your changes tomorrow.

**Review by lukmccall — CHANGES_REQUESTED** [review]
Looking good 🏅
I apologize for the delayed response, as I have been quite occupied lately.
Please, apply my suggestions then I will merge your pr.

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/scanners/MLKitBarCodeScanner.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  addOnCanceledListener {
+    continuation.cancel()
+  }
+}
+
+class MLKitBarCodeScanner(context: Context) : ExpoBarCodeScanner(context) {
+
+  private var barcodeScannerOptions =
+    BarcodeScannerOptions.Builder()
+      .setBarcodeFormats(Barcode.FORMAT_ALL_FORMATS) // Barcode.FORMAT_ALL_FORMATS
```
[the remark this answers is not in the document]
Comment (toshiyuki-suzuki-yukashikado) [reply packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/scanners/MLKitBarCodeScanner.kt]: Comment out was removed.
[1511e1a](https://github.com/expo/expo/pull/22107/commits/1511e1ab408d33fbf3537aace13249dc5bbf376e)

File: packages/expo-barcode-scanner/android/build.gradle [hunk 2]
```diff
@@ -85,6 +85,7 @@ dependencies {
   implementation project(':expo-modules-core')
 
   api 'com.google.android.gms:play-services-vision:19.0.0'
```
[the remark this answers is not in the document]
Comment (toshiyuki-suzuki-yukashikado) [reply packages/expo-barcode-scanner/android/build.gradle]: Yes, we can.  `play-services-vision` doesn't depend on anything. 

[7afd682](https://github.com/expo/expo/pull/22107/commits/7afd682e35cbb0552cac2a86b08bf583a154479d)

**Comment by toshiyuki-suzuki-yukashikado** [comment]
@lukmccall 
Thank you for reviewing!  And, sorry for the late fix.  I fixed your suggestions.
Please review again!

**Review by tsapeta — APPROVED** [review]
Nice, this looks great to me! Thank you for your contribution. I only left a suggestion regarding the changelog entry, otherwise it's ready to go 🤩

**Comment by toshiyuki-suzuki-yukashikado** [comment]
@tsapeta 
Thank you for reviewing! 
I should have write the chargelog descriptive. I fixed it!

@lukmccall
Can this PR be merged?

**Comment by tsapeta** [comment]
@toshiyuki-suzuki-yukashikado thanks for applying these suggestions. Looks like some CI jobs are failing, see https://github.com/expo/expo/actions/runs/5173125149/jobs/9324047108#step:10:1597
Could you take a look why this is failing?

File: packages/expo-barcode-scanner/android/build.gradle [hunk 3]
```diff
@@ -81,8 +81,7 @@ android {
 
 dependencies {
   implementation project(':expo-modules-core')
-
-  api 'com.google.android.gms:play-services-vision:19.0.0'
+  api 'com.google.mlkit:barcode-scanning:17.1.0'
   api 'com.google.zxing:core:3.3.3'
```
[the remark this answers is not in the document]
Comment (toshiyuki-suzuki-yukashikado) [reply packages/expo-barcode-scanner/android/build.gradle]: Yes! Zxing was removed in 31b10c0

**Comment by toshiyuki-suzuki-yukashikado** [comment]
@lukmccall
Only ios test is failed. I edited only android code, not ios code. 
Is it possible to merge with ios test errors? Please check it🙏

**Review by tsapeta — APPROVED** [review]

File: packages/expo-barcode-scanner/android/build.gradle [hunk 4]
```diff
@@ -81,10 +81,7 @@ android {
 
 dependencies {
   implementation project(':expo-modules-core')
-
-  api 'com.google.android.gms:play-services-vision:19.0.0'
-  api 'com.google.zxing:core:3.3.3'
-
+  api 'com.google.mlkit:barcode-scanning:17.1.0'
```
Comment (younes0) [inline packages/expo-barcode-scanner/android/build.gradle:84]: this adds `libbarhopper_v3.so` which dramatically increases the build size
see: https://github.com/expo/expo/issues/26749

**Pull request merged**